### PR TITLE
RFC: Cache trait

### PR DESCRIPTION
**Pull request by bendavies**
| Q             | A
| ------------- | ---
| Bug fix?      | no
| New feature?  | no
| BC breaks?    | no
| Deprecations? | no
| Tests pass?   | yes
| Fixed tickets | 
| License       | MIT
| Doc PR        | 

cc @dunglas  as discussed.

I noticed we were doing the same thing in quite a few places.
Thoughts please?

P.S. i've put very little thought into the naming of the trait and function. Please suggest something better!

### CODE REVIEW

**Comment by teohhanhui** [comment]
-1 as it's unnecessary and only makes the code more confusing.

**Comment by dunglas** [comment]
I'm +1 to factorize this code, it's an error prone one, better not duplicate it everywhere.

**Comment by bendavies** [comment]
@dunglas preference on trait vs abstract class?

**Comment by teohhanhui** [comment]
In that case I'm -1 for something confusing to the reader of the code like `doAOrB`, in this case `getOrSave`.

**Comment by bendavies** [comment]
@teohhanhui did you even read the PR description?

**Comment by teohhanhui** [comment]
Oh, I don't mean -1 for this PR. Sorry for the confusion.

File: src/Bridge/Symfony/Routing/CachedRouteNameResolver.php [hunk 1]
```diff
@@ -43,33 +44,8 @@ public function getRouteName(string $resourceClass, $operationType /**, array $c
...
-
-        try {
-            $cacheItem->set($routeName);
-            $this->cacheItemPool->save($cacheItem);
-        } catch (CacheException $e) {
-            // do nothing
-        }
-
-        return $this->localCache[$cacheKey] = $routeName;
+        return $this->getOrSave($cacheKey, $this->cacheItemPool, function () use ($resourceClass, $operationType, $context) {
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/CachedRouteNameResolver.php]: Perhaps `getCached`? I also agree with @soyuka that we shouldn't need to pass the `$cacheItemPool` here, it should be in the trait.

File: src/Bridge/Symfony/Routing/CachedRouteNameResolver.php [hunk 2]
```diff
@@ -13,7 +13,7 @@
 
 namespace ApiPlatform\Core\Bridge\Symfony\Routing;
 
-use Psr\Cache\CacheException;
+use ApiPlatform\Core\Cache\CacheTrait;
```
Comment (teohhanhui) [inline src/Bridge/Symfony/Routing/CachedRouteNameResolver.php]: `CachedTrait`?

File: src/Cache/CacheTrait.php [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+namespace ApiPlatform\Core\Cache;
+
+use Psr\Cache\CacheException;
+use Psr\Cache\CacheItemPoolInterface;
+
+trait CacheTrait
+{
+    private $localCache = [];
+
+    private function getOrSave(string $cacheKey, CacheItemPoolInterface $pool, callable $callback)
```
Comment (soyuka) [inline src/Cache/CacheTrait.php]: naming proposal:

```php
<?php
private function cache(string $cacheKey, callable $getValue) {

}
```

With a parent class we can remove CacheItemPoolInterface from the method signature

Comment (teohhanhui) [reply src/Cache/CacheTrait.php]: We don't need a parent class. The property could be declared in the trait as we already do elsewhere?

File: src/Cache/CacheTrait.php [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+ */
+
+declare(strict_types=1);
+
+namespace ApiPlatform\Core\Cache;
+
+use Psr\Cache\CacheException;
+use Psr\Cache\CacheItemPoolInterface;
+
+trait CacheTrait
```
Comment (dunglas) [inline src/Cache/CacheTrait.php]: Can you make it `@internal`?

File: src/Cache/CacheTrait.php (hunk 3 above)
Comment (soyuka) [reply src/Cache/CacheTrait.php]: also works yes, I just think that parent class > trait if we can :) (ability to use `instanceof parent` etc.)

Comment (teohhanhui) [reply src/Cache/CacheTrait.php]: I think using inheritance here is... wrong. It's the kind of use case which traits are exactly designed to solve - horizontal code reuse.

**Comment by bendavies** [comment]
comments addressed.
Note that has this has not been added to `CachedIdentifiersExtractor` as it looks a little more complicated, but might be possible.

File: src/Cache/CachedTrait.php [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+ */
+trait CachedTrait
+{
+    /** @var CacheItemPoolInterface */
+    private $cacheItemPool;
+    private $localCache = [];
+
+    private function getCached(string $cacheKey, callable $getValue)
+    {
+        if (isset($this->localCache[$cacheKey])) {
```
Comment (teohhanhui) [inline src/Cache/CachedTrait.php]: Actually, I think this is a bug. We should support storing `null` values in the cache.

**Comment by teohhanhui** [comment]
Thanks @bendavies! :tada:

**Pull request merged**